mantle/cmd/ore/ibmcloud/copy-object.go

```diff
@@ -0,0 +1,77 @@
+// Copyright 2021 RedHat.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ibmcloud
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	cmdCopyObject = &cobra.Command{
+		Use:   "copy-object <dest-bucket...>",
+		Short: "Copy IBMCloud object to a bucket",
+		Long:  "Copy an IBMCloud object from a bucket in a region to another bucket in a specified region",
+		RunE:  runCopyObject,
+
+		SilenceUsage: true,
+	}
+
+	copyCloudObjectStorage string
+	sourceBucket           string
+	sourceName             string
+	destRegion             string
+	destBucket             string
+)
+
+func init() {
+	IbmCloud.AddCommand(cmdCopyObject)
+	cmdCopyObject.Flags().StringVar(&copyCloudObjectStorage, "cloud-object-storage", "coreos-dev-image-ibmcloud", "cloud object storage to be used")
+	cmdCopyObject.Flags().StringVar(&sourceBucket, "source-bucket", "coreos-dev-image-ibmcloud-us-east", "bucket where object needs to be copied from")
+	cmdCopyObject.Flags().StringVar(&sourceName, "source-name", "", "name of object to be copied")
+	cmdCopyObject.Flags().StringVar(&destRegion, "destination-region", "", "region to be copied to")
+	cmdCopyObject.Flags().StringVar(&destBucket, "destination-bucket", "", "destination bucket to copy to")
+}
+
+func runCopyObject(cmd *cobra.Command, args []string) error {
+	if sourceName == "" || destRegion == "" || destBucket == "" {
+		fmt.Fprintf(os.Stderr, "Specify the object name, destination region and bucket.\n")
+		os.Exit(2)
+	}
+
+	if err := API.NewS3Client(copyCloudObjectStorage, destRegion); err != nil {
+		return err
+	}
+
+	bucketExists, err := API.CheckBucketExists(destBucket)
+	if err != nil {
+		return err
+	}
+
+	if !bucketExists {
+		if err = API.CreateBucket(destBucket); err != nil {
+			return err
+		}
+	}
```

Comment:
IMO, bucket creation in such automation is always risky, I prefer to have a predefined bucket created out-of-bound to be a little flexible.

This is because:
- This will avoid if the same bucket name exist else where
- We can have only object creation auth to the API_KEY to restrict to certain operations only(better security model)

Reply:
absolutely. for the pipeline, the buckets will already be created. this is just for completeness that i added it because this command can be used even outside the pipeline.